joy_teleop/joy_teleop/joy_teleop.py

```diff
@@ -251,71 +229,167 @@ def run_topic(self, c, joy_state):
+
+        self.commands = []
+
+        names = []
+
+        for name, config in self.retrieve_config().items():
+            if name in names:
+                self.get_logger().error("command '{}' was duplicated".format(name))
+                raise JoyTeleopException("command '{}' was duplicated".format(name))
+
             try:
-                package, interface, message = type_name.split('/')
-                mod = importlib.import_module(package + ext)
-                self.message_types[type_name] = getattr(mod, message)
-            except ValueError:
-                raise JoyTeleopException('message type format error')
-            except ImportError:
-                raise JoyTeleopException('module {} could not be loaded'.format(package))
-            except AttributeError:
-                raise JoyTeleopException(
-                    'message {} could not be loaded from module {}'.format(package, message))
-        return self.message_types[type_name]
-
-    def update_actions(self, evt=None):
-        for name, cmd in self.command_list.items():
-            if cmd['type'] != 'action':
-                continue
-            if cmd['action_name'] in self.offline_actions:
-                self.register_action(name, cmd)
+                interface_group = config['type']
+
+                if interface_group == 'topic':
+                    self.commands.append(JoyTeleopTopicCommand(name, config, self))
+                elif interface_group == 'service':
+                    self.commands.append(JoyTeleopServiceCommand(name, config, self))
+                elif interface_group == 'action':
+                    self.commands.append(JoyTeleopActionCommand(name, config, self))
+                else:
+                    self.get_logger().error("unknown type '{interface_group}' "
+                                            "for command '{name}'".format_map(locals()))
+                    raise JoyTeleopException("unknown type '{interface_group}' "
+                                             "for command '{name}'".format_map(locals()))
+            except TypeError:
+                # This can happen on parameters we don't control, like
+                # 'use_sim_time'.  Just ignore it with a warning for the user.
+                self.get_logger().warn('parameter {} is not a dict'.format(name))
+
+            names.append(name)
+
+        # Don't subscribe until everything has been initialized.
+        self._subscription = self.create_subscription(
+            sensor_msgs.msg.Joy, 'joy', self.joy_callback, 1)
+
+    def retrieve_config(self):
+        config = {}
+        for param_name in sorted(self._parameters.keys()):
+            pval = self.get_parameter(param_name).value
+            self.insert_dict(config, param_name, pval)
+        return config
+
+    def insert_dict(self, dictionary, key, value):
+        split = key.split(PARAMETER_SEPARATOR_STRING, 1)
```

Comment:
@clalancette nit^N: there's [`str.partition`](https://docs.python.org/3.7/library/stdtypes.html#str.partition) too

Reply:
Hm, true.  I don't know that one is more readable than the other in this case.  Is there an additional benefit to using `str.partition`?

Reply:
I'd say that API is there for exactly this use case, but otherwise both should be equally performant (unless a `tuple` is objectively cheaper to instantiate than a `list`).

Reply:
Same about this applying, but won't block on it.